### PR TITLE
[Funk] New extractor

### PR DESCRIPTION
**Pull request by catdog2**
## Please follow the guide below
- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your _pull request_ (like that [x])
- Use _Preview_ tab to see how your _pull request_ will actually look like

---
### Before submitting a _pull request_ make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)
### What is the purpose of your _pull request_?
- [x] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---
### Description of your _pull request_ and other information

This adds an extractor for https://www.funk.net/, a new site made by the German public broadcasting TV stations. This also fixes a small bug in the  Kaltura extractor on which the new extractor relies on.

### CODE REVIEW

File: youtube_dl/extractor/funk.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            }
+        }
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        title = self._html_search_regex(r'<title>(.+?)</title>', webpage,
+                                        'title', fatal=False)
+        # remove the | funk
```
Comment (yan12125) [inline youtube_dl/extractor/funk.py]: Use ```remove_end``` instead in this case.

File: youtube_dl/extractor/funk.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        title = title.rsplit('|', 1)[0].strip()
+
+        description = self._html_search_regex(
+            r'<meta\s+name="description"\s+content="([^"]+)"',
+            webpage, "description", fatal=False)
```
Comment (yan12125) [inline youtube_dl/extractor/funk.py]: Use ```_html_search_meta``` instead

**Comment by jlnostr** [comment]
Any update on this? @catdog2

**Comment by catdog2** [comment]
The code also now uses `_html_search_meta` and  `remove_end`. In addition I noted in the commit message that it needs more work to also support https://www.funk.net/serien/...

**Comment by jlnostr** [comment]
Nice 👍

I did some investigating for the `serien` support. First of all, you need to add the `Referer: https://www.funk.net` HTTP header for each request. This will at least validate the *.m3u download request. In the case of the trailer for ['Hoff the Record'](https://www.funk.net/serien/57e7de0de4b0a1795f1480e8/items/57ee7535e4b078785fe8a919) this will lead you to the following playlist:

```
#EXTM3U
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=855000,RESOLUTION=640x360,CODECS="avc1.77.30, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/index_0_av.m3u8?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=%2fi%2f20160930%2f1985%2f1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=1441000,RESOLUTION=1280x720,CODECS="avc1.77.30, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/index_1_av.m3u8?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=%2fi%2f20160930%2f1985%2f1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=2320000,RESOLUTION=1280x720,CODECS="avc1.640028, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/index_2_av.m3u8?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=%2fi%2f20160930%2f1985%2f1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXT-X-STREAM-INF:PROGRAM-ID=1,BANDWIDTH=3648000,RESOLUTION=1920x1080,CODECS="avc1.64001f, mp4a.40.2",CLOSED-CAPTIONS=NONE
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/index_3_av.m3u8?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=%2fi%2f20160930%2f1985%2f1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
```

The file behind the 1080p string looks like this:
```
#EXTM3U
#EXT-X-TARGETDURATION:10
#EXT-X-ALLOW-CACHE:YES
#EXT-X-KEY:METHOD=AES-128,URI="https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/crypt.key?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=/i/20160930/1985/1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12"
#EXT-X-PLAYLIST-TYPE:VOD
#EXT-X-VERSION:3
#EXT-X-MEDIA-SEQUENCE:1
#EXTINF:10.000,
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/segment1_3_av.ts?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=/i/20160930/1985/1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXTINF:10.000,
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/segment2_3_av.ts?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=/i/20160930/1985/1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXTINF:10.000,
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/segment3_3_av.ts?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=/i/20160930/1985/1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXTINF:10.000,
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/segment4_3_av.ts?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=/i/20160930/1985/1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXTINF:10.000,
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/segment5_3_av.ts?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=/i/20160930/1985/1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXTINF:10.000,
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/segment6_3_av.ts?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=/i/20160930/1985/1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXTINF:5.646,
https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/segment7_3_av.ts?null=0&id=AgBP0tMEF0oYd%2fsSBliEwXkTYzKd9yhx6BDF2m7w%2fbWSl8p3UQ6N5M%2ftHI9GLrzoSpGnrx2p9HeLIg%3d%3d&hdntl=exp=1476879483~acl=/i/20160930/1985/1_84je45no_1_*~data=hdntl~hmac=ee9eff7c52725c3782ced2687e885e0455258e54a72dfd443e05ee52c060eb12
#EXT-X-ENDLIST
```

As you can see it's crypted. Can `youtube-dl` handle this and if so, is there a way to extract a direct download link from the above playlist?

**Comment by Crypt0Graf** [comment]
@jlnostr Hi, I've tried your way, but it didn't work. You've probably tried it with the answers/content of the master.m3u8, but it didn't work. However, youtube-dl works fine with the simple link of the master.m3u8. You can try this with `youtube-dl [playlist link]`.
I'm not sure how restricted this link is in concern of token and validity, but in my case `youtube-dl https://jaondemandsecure-vh.akamaihd.net/i/20160930/1985/1_84je45no_1_,vjrttms0,3hh61r8k,yh13ft6h,i2m6qup1,_1.mp4.csmil/master.m3u8?hdnts=st=1477511460~exp=1477516860~acl=/i/20160930/1985/1_84je45no_1_*~hmac=e53f319a51a76f176235970380ba1d731d9af8e65f8b82e6af7dad545b152382 -o outputname` worked fine. 
So yes, youtube-dl can handle funk.net videos, the extractor for that just isn't implemented yet.

**Comment by Boris-de** [comment]
I added support for "serien" and the overview-pages (as playlists) to support downloading of a whole series. You can find the change in my [funk](https://github.com/Boris-de/youtube-dl/tree/funk)-branch: Boris-de@ef5ac4610347d7a83fef4bf226951c605f402c54

Tested with two series, both worked fine. I added "Wishlist" to the tests.

@catdog2: Feel free to just amend my changes to your pull request.

**Comment by jlnostr** [comment]
@Boris-de You're the hero of my day 👍

**Comment by jlnostr** [comment]
I've did some testing today with the code from @Boris-de and everything works like expected. In my opinion the code is ready for production. It would be nice if @catdog2 could amend the changes (maybe @Boris-de needs to create a PR to his branch) :)

**Comment by jlnostr** [comment]
One issue I found is the download of FSK 16 series, like Outcast. I can't download them, even after 10pm (GMT+1).

**Pull request closed without merging**